### ISSUE TITLE
[Feature request] - Support kitty terminal's graphics protocol

### ISSUE THREAD

**Issue by Strykar** [weechat/weechat#1430]
Consider supporting kitty terminal's graphic protocol.
This would enable viewing images inside the terminal (kitty) that runs weechat and thus in weechat.

Reasons for supporting it over current (_image in terminal_) implementations are discussed at https://github.com/kovidgoyal/kitty/issues/33

Protocol document (1 page long) - https://sw.kovidgoyal.net/kitty/graphics-protocol.html

Also see - https://github.com/kovidgoyal/kitty/issues/2184

**Comment by virus-found** [comment]
Or support `libsixel` to make it usable on a range of terminals, not just kitty.

**Comment by Strykar** [comment]
> Or support `libsixel` to make it usable on a range of terminals, not just kitty.

Arguments on sixel being an inferior solution - https://github.com/kovidgoyal/kitty/issues/33

**Comment by DrSensor** [comment]
Fyi, [wezterm](https://wezfurlong.org/wezterm/imgcat.html) (a cross-platform terminal emulator) support iTerm2 inline images protocol.

They library called [termwiz](https://docs.rs/termwiz/latest/termwiz/caps/struct.Capabilities.html) has a helper to detect terminal capabilities, including whatever the running terminal support image preview or not.

**Comment by kidonng** [comment]
> The library they created called [termwiz](https://docs.rs/termwiz/latest/termwiz/caps/struct.Capabilities.html) has a helper to detect terminal capabilities, including whatever the running terminal support image preview or not and which protocol it support.

There is no magic happening, it just [checks whether the terminal is iTerm or WezTerm](https://github.com/wez/wezterm/blob/a59e9b1706be78beb722c7c6362687e13c0ec30b/termwiz/src/caps/mod.rs#L254-L271).

**Comment by r3k2** [comment]
I will switch back to weechat from gomuks in a snap if there is sixel or any other graphics support.

**Comment by narodnik** [comment]
would love sixel or any image support so we can share math equations rendered from latex. that would be amazing for our work, or even shitposting memes. everyone today is using normie chat but lets bring back terminal UI and IRC!

**Comment by trygveaa** [comment]
For WeeChat to add support for graphics, ncurses would need support for it first (unless WeeChat is ported to another UI library), and I don't see that happening any time soon.

You can technically signal the terminal to display an image on top of WeeChat, but it wouldn't follow scrolling, buffer changes etc.

However, there is some discussion around extending kittys graphics protocol to place images using unicode characters, which would allow it to work in any program that can pass escape sequences to the terminal and display unicode characters. I've tested it and it works in WeeChat, including properly following scrolling, buffer changes etc. You can read about it here: https://github.com/kovidgoyal/kitty/discussions/4021

**Comment by narodnik** [comment]
ncurses can work with sixel graphics:

https://lists.gnu.org/archive/html/bug-ncurses/2016-12/msg00022.html

There's even a file in the repo called sixel.c:

https://github.com/dankamongmen/notcurses/blob/master/src/lib/sixel.c

**Comment by trygveaa** [comment]
> ncurses can work with sixel graphics:
> 
> https://lists.gnu.org/archive/html/bug-ncurses/2016-12/msg00022.html

That mail describes just displaying images by bypassing ncurses. That's what I meant by:

> You can technically signal the terminal to display an image on top of WeeChat, but it wouldn't follow scrolling, buffer changes etc.

For that to be useful in WeeChat, you'd have to set aside an area for images, e.g. create a bar of a certain size that you don't have any bar items in, and only display images there.

> There's even an ncurses derived library with sixel support:
> 
> https://github.com/dankamongmen/notcurses/blob/master/src/lib/sixel.c

Notcurses is not ncurses derived, it's an alternative to ncurses, so WeeChat would have to be ported to it. I've thought about it, but it would be a huge effort.

**Comment by Eragonfr** [comment]
Could it be possible to do image preview in a plugin instead of in weechat ?

**Comment by trygveaa** [comment]
The Kitty graphics protocol now has support for displaying images within terminal programs with no special support from the terminal program (except for unicode, diacritics, text color and the ability to send escape codes to the terminal, all of which WeeChat supports). Using this I made a script to display images in WeeChat: https://github.com/trygveaa/weechat-icat (currently requires a development version of kitty, the support will be included in the next release of kitty).

**Comment by flashcode** [comment]
@trygveaa: do you have a screenshot of the result? It might be interesting to have this in your README.

**Comment by trygveaa** [comment]
Sure, here's a screenshot. I'll put it in the readme as well.

![weechat-icat-demo](https://user-images.githubusercontent.com/601966/224491522-5766c7af-97cb-4389-93d2-d992566229a2.png)

You can't see it from just a screenshot, but it's worth noting that the images will always be in the correct position, so they scroll just like other text and are only shown in the buffer you display them in. Only thing that acts weird is if you resize the window to be narrower than the image. The image won't resize and will look weird until you make the window larger again.

**Comment by schrmh** [comment]
> For WeeChat to add support for graphics, ncurses would need support for it first (unless WeeChat is ported to another UI library), and I don't see that happening any time soon.

@ThomasDickey your input here should be very valuable.
(Thomas E. Dickey maintains e.g. ncurses and xterm; the latter can be compiled with SIXEL support)

**Comment by tempodat** [comment]
Now that Microsoft Terminal has added sixel support, I suggest refloating this idea. This could be the vehicle for the modernization of CLI IRC, and CLI Interfaces in general. Leading is better than being left out.

**Comment by aiobofh** [comment]
Yes. Sixels is the way.

**Comment by ManuGraiph** [comment]
I'm a bit late to the party; my 2-cents:

If the point of this was to display received images, with the `xfer_run_command` plugin and a bit of scripting you can pretty much do anything; you just have to write a simple script that opens the image as a floating window, fetch the coordinates/ws of weechat, and voilá.

**Comment by Strykar** [comment]
I audited this: read the WeeChat curses source and the code of @trygveaa's `weechat-icat`. Findings:

**Effort.**
A native in-core implementation is a multi-week job and mostly redundant, because the script already does the hard part correctly via the only mechanism that can work. The work actually worth doing is small (bottom).

**Why in-core graphics is hard, from the source.**
WeeChat flushes the screen through ncurses `refresh()` only (there is no `doupdate` seam anywhere in src/gui/curses), it re-renders the whole chat rectangle on every printed line, and it destroys and recreates the chat window on buffer switch and on resize. So anything drawn to the terminal that ncurses does not know about is overwritten on the next paint. That rules out cursor-placed images and sixel for an inline scrolling chat: the pixmap desyncs the moment text scrolls.

**Why unicode placeholders are the only approach that survives, and need no core change.**
In `weechat-icat` the image bytes are written out of band straight to the terminal device, never through WeeChat. The on-screen placement is just unicode placeholder characters plus an 8-bit color (plus diacritics) printed through the normal print path, so ncurses stores, scrolls and repaints them like any glyph. Two things people assume block a plugin turn out not to be needed: you do not need to push raw escapes through printed text (that was #2058, closed won't fix), and you do not need truecolor (the image id rides in an 8-bit color).

**What is actually missing is convenience, not capability.**
`weechat-icat` is a manual `/icat <file>`. There is no auto preview of image links posted in a channel. That is a small script on top of what exists: detect the URL, download it (WeeChat's `url:` hook), call the existing display path. The one rough edge is that an image does not reflow if you shrink the window narrower than it.

Sixel (no kitty needed) is a separate request, #2304, and the ncurses maintainer has said it is not planned for ncurses.

TLDR: Achieving this (Kitty's graphic protocol or Sixel) would require upstream changes in ncurses which the devs are not open to and the practical issue is largely solved by https://github.com/trygveaa/weechat-icat
